### PR TITLE
[WIP] Added support for a variety of SSH clients, notably putty

### PR DESCRIPTION
**Pull request by robeving**
Thank you for a great library!

I've added support for:
* password based authentication.
* the creation of pseudo terminals - this should also fix issue #3 .
* event handlers to support the above.

I've also fixed a number of bugs.

### CODE REVIEW

**Comment by robeving** [comment]
Wow, two PRs for the same issue in the same day.

I've looked at the other PR, it doesn't contain as much functionality as this one. Such as a few new message types and handlers to allow you to build your own terminal.

I've also tested this PR quite heavily.

File: FxSsh/Messages/Connection/DirectTcpIpMessage.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+            base.OnLoad(reader);
+
+            if (ChannelType != "direct-tcpip")
+                throw new ArgumentException(string.Format("Channel type {0} is not valid.", ChannelType));
+
+            Host = reader.ReadString(Encoding.ASCII);
+            Port = reader.ReadUInt32();
+            OriginatorIPAddress = reader.ReadString(Encoding.ASCII);
+            OriginatorPort = reader.ReadUInt32();
+
```
Comment (Aimeast) [inline FxSsh/Messages/Connection/DirectTcpIpMessage.cs]: remove an empty line

File: FxSsh/Messages/Connection/EnvMessage.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿using FxSsh.Messages.Connection;
+using System.Text;
+
+namespace FxSsh.Messages
+{
+    public class EnvMessage : ChannelRequestMessage
+    {
+        public string Name = "";
```
Comment (Aimeast) [inline FxSsh/Messages/Connection/EnvMessage.cs]: use property
```suggestion
        public string Name {get;set;}        
```

File: FxSsh/Messages/Connection/ForwardedTcpIpMessage.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            base.OnLoad(reader);
+
+            if (ChannelType != "forwarded-tcpip")
+                throw new ArgumentException(string.Format("Channel type {0} is not valid.", ChannelType));
+
+            Address = reader.ReadString(Encoding.ASCII);
+            Port = reader.ReadUInt32();
+            OriginatorIPAddress = reader.ReadString(Encoding.ASCII);
+            OriginatorPort = reader.ReadUInt32();
+
```
Comment (Aimeast) [inline FxSsh/Messages/Connection/ForwardedTcpIpMessage.cs]: empty line

File: FxSsh/Messages/Connection/ShellRequestMessage.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Text;
+
+namespace FxSsh.Messages.Connection
+{
+    public class ShellRequestMessage : ChannelRequestMessage
+    {
+        protected override void OnLoad(SshDataWorker reader)
```
Comment (Aimeast) [inline FxSsh/Messages/Connection/ShellRequestMessage.cs]: remove unused override

File: FxSsh/Messages/Connection/ShouldIgnoreMessage.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+namespace FxSsh.Messages.Connection
+{
+    [Message("SSH_MSG_IGNORE", MessageNumber)]
+    public class ShouldIgnoreMessage : ConnectionServiceMessage
+    {
+        private const byte MessageNumber = 2;
+
+        public override byte MessageType { get { return MessageNumber; } }
+
+        protected override void OnLoad(SshDataWorker reader)
```
Comment (Aimeast) [inline FxSsh/Messages/Connection/ShouldIgnoreMessage.cs:11]: remove unused override

File: FxSsh/Messages/Connection/SubsystemMessage.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿using FxSsh.Messages.Connection;
+using System.Text;
+
+namespace FxSsh.Messages
+{
+    public class SubsystemMessage : ChannelRequestMessage
+    {
+        public string Name = "";
```
Comment (Aimeast) [inline FxSsh/Messages/Connection/SubsystemMessage.cs]: use property

File: FxSsh/Services/UserauthArgs.cs [hunk 7]
```diff
@@ -4,7 +4,7 @@ namespace FxSsh.Services
 {
     public class UserauthArgs
     {
-        public UserauthArgs(string keyAlgorithm, string fingerprint, byte[] key)
+        public UserauthArgs(Session s, string username, string keyAlgorithm, string fingerprint, byte[] key)
```
Comment (Aimeast) [inline FxSsh/Services/UserauthArgs.cs]: rename `s` as `session`
```suggestion
        public UserauthArgs(Session session, string username, string keyAlgorithm, string fingerprint, byte[] key)
```

File: FxSsh/Services/UserauthService.cs [hunk 8]
```diff
@@ -35,33 +36,76 @@ private void HandleMessage(RequestMessage message)
             switch (message.MethodName)
             {
                 case "publickey":
-                    var msg = Message.LoadFrom<PublicKeyRequestMessage>(message);
-                    HandleMessage(msg);
+                    var keyMsg = Message.LoadFrom<PublicKeyRequestMessage>(message);
+                    HandleMessage(keyMsg);
```
Comment (Aimeast) [inline FxSsh/Services/UserauthService.cs]: trim the white space at end
```suggestion
                    HandleMessage(keyMsg);
```

File: FxSsh/Services/UserauthService.cs [hunk 9]
```diff
@@ -35,33 +36,76 @@ private void HandleMessage(RequestMessage message)
             switch (message.MethodName)
             {
                 case "publickey":
-                    var msg = Message.LoadFrom<PublicKeyRequestMessage>(message);
-                    HandleMessage(msg);
+                    var keyMsg = Message.LoadFrom<PublicKeyRequestMessage>(message);
+                    HandleMessage(keyMsg);       
                     break;
                 case "password":
+                    var msg = Message.LoadFrom<PasswordRequestMessage>(message);
```
Comment (Aimeast) [inline FxSsh/Services/UserauthService.cs]:
```suggestion
                    var pswdMsg = Message.LoadFrom<PasswordRequestMessage>(message);
                    HandleMessage(pswdMsg);
                    break;
```

File: FxSsh/Services/UserauthService.cs [hunk 10]
```diff
@@ -71,28 +115,13 @@ private void HandleMessage(PublicKeyRequestMessage message)
...
                     if (Succeed != null)
-                        Succeed(this, message.ServiceName);
-                    _session.SendMessage(new SuccessMessage());
-                    return;
-                }
-                else
                 {
-                    _session.SendMessage(new FailureMessage());
-                    throw new SshConnectionException("Authentication fail.", DisconnectReason.NoMoreAuthMethodsAvailable);
+                    Succeed(this, message.ServiceName);
```
Comment (Aimeast) [inline FxSsh/Services/UserauthService.cs]: must check `Succeed` is not null

File: FxSsh/Services/UserauthService.cs [hunk 11]
```diff
@@ -35,33 +36,76 @@ private void HandleMessage(RequestMessage message)
...
+                Userauth(this, args);
+                verifed = args.Result;
+            }
+
+            if (verifed)
+            {
+                _session.RegisterService(message.ServiceName, args);
+
+                if (Succeed != null)
+                    Succeed(this, message.ServiceName);
```
Comment (Aimeast) [inline FxSsh/Services/UserauthService.cs]: I'd like use new grammar
```suggestion
                Succeed.?(this, message.ServiceName);
```

**Comment by robeving** [comment]
I've updated the PR based on your comments

**Comment by Aimeast** [comment]
Thanks for your cooperation, @robeving . It is amazing!

I have only reviewed the styles of your code, the logic are not reviewed yet. Could you provide your test or sample code for your local test, because I can't know how handle the new features below.

```cs
        public event EventHandler<SessionRequestedArgs> CommandOpened;
        public event EventHandler<EnvironmentArgs> EnvReceived;
        public event EventHandler<PtyArgs> PtyReceived;
        public event EventHandler<TcpRequestArgs> TcpForwardRequest;
        public event EventHandler<TcpDataArgs> TcpData;
```

**Comment by robeving** [comment]
I've updated the SshServerLoader to include an example of how the new event handlers can be used.

**Comment by robeving** [comment]
I've also updated the readme to include the sample code

**Comment by robeving** [comment]
I've applied your patch & updated the example code with how to send a response.

In regard to your other comments I call CommandOpened when a shell/subsystem is requested because if I didn't you wouldn't be able to registered for other events in time.  IIRC you don't get a CommandRequestMessage in these cases, or at least i tested it with a few SSH clients and didn't see it.

As for port forwarding closing the channel is not ideal. My current use case is HTTP which handles this correctly. I'm working on other protocols in the next 6 months.

**Comment by robeving** [comment]
BUMP

**Comment by Aimeast** [comment]
I add the mark [WIP]. there are still more works need to do. As my ideal about port forwarding, a message loop or other async methods (IOCP/SAEA) to be created after open a forwarding channel, in order to receive and forward message between two endpoint.

**Comment by robeving** [comment]
I've updated the code. The example gives a good idea on how it could be used to provide an end-to-end proxying service.

I've tested it with a few protocols where the server / client issues data first.

**Comment by Aimeast** [comment]
I am still testing the port forwarding. looks like there are more difficulty going to test.

**Comment by robeving** [comment]
I've been using this in production for a while without issue. What problems do you have testing?

**Comment by watsonsong** [comment]
Hi @robeving, thanks for this support. I am using the code and test in my project, and hope the putty support can be update so I can dependent on the nuget.

**Comment by Aimeast** [comment]
A `Incorrect MAC received on packet` occur during PuTTY session. But I think we can merge this PR and open a new issue to follow up.

@robeving Please rebase to master and squash into one commit I will merge it.

**Pull request closed without merging**